### PR TITLE
8279930: Synthetic cast causes generation of store barriers when using heap segments

### PR DESCRIPTION
**Pull request by mcimadamore**
When looking at larger benchmarks, I noted a discrepancy between the performance of off-heap segments and on-heap segments. Looking at the assembly for the `MemorySegment::asSlice` method I could see some additional barriers in the off-heap case, but could not initially make sense of them. Vlad pointed me at G1 (in fact no such barrier was emitted when using a different GC, such as the serial GC, or ZGC), and later Erik narrowed the problem down to a failure in a C2 optimization to remove barriers around initializing stores. This problem was caused by a synthetic cast added by javac to a value (the base object) that initialized the newly created memory segment slice. Because of that case, C2 missed the store as an "initializing" one, and inserted additional barriers. This patch should make performance of on-heap segments a lot more reliable, especially when slicing is involved.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8279930](https://bugs.openjdk.java.net/browse/JDK-8279930): Synthetic cast causes generation of store barriers when using heap segments

### Reviewers
 * [Paul Sandoz](https://openjdk.java.net/census#psandoz) (@PaulSandoz - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk18 pull/97/head:pull/97` \
`$ git checkout pull/97`

Update a local copy of the PR: \
`$ git checkout pull/97` \
`$ git pull https://git.openjdk.java.net/jdk18 pull/97/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 97`

View PR using the GUI difftool: \
`$ git pr show -t 97`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk18/pull/97.diff">https://git.openjdk.java.net/jdk18/pull/97.diff</a>

</details>

### CODE REVIEW

File: src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/HeapMemorySegmentImpl.java [hunk 1]
```diff
@@ -84,9 +84,9 @@ ByteBuffer makeByteBuffer() {
 
     // factories
 
-    public static class OfByte extends HeapMemorySegmentImpl<byte[]> {
+    public static class OfByte extends HeapMemorySegmentImpl {
 
-        OfByte(long offset, byte[] base, long length, int mask) {
+        OfByte(long offset, Object base, long length, int mask) {
```
Comment (PaulSandoz) [inline src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/HeapMemorySegmentImpl.java:93]: Is it possible to retain the array type for the constructor?

Comment (mcimadamore) [reply src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/HeapMemorySegmentImpl.java:93]: yes, that should be harmless

Comment (mcimadamore) [reply src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/HeapMemorySegmentImpl.java:93]: Actually, no, sorry. If I sharpen the type, then `dup` has to cast, so we go back to the original issue.

Comment (PaulSandoz) [reply src/jdk.incubator.foreign/share/classes/jdk/internal/foreign/HeapMemorySegmentImpl.java:93]: Ah yes, i see. Maybe it's worth adding a general comment to the class doc of `HeapMemorySegmentImpl` briefly explaining the reasons for use of `Object`.

File: test/micro/org/openjdk/bench/jdk/incubator/foreign/LoopOverSlice.java [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+    static final int ALLOC_SIZE = ELEM_SIZE * CARRIER_SIZE;
+
+    MemorySegment nativeSegment, heapSegment;
+    ResourceScope scope;
+
+    @Setup
+    public void setup() {
+        scope = ResourceScope.newConfinedScope();
+        nativeSegment = MemorySegment.allocateNative(ALLOC_SIZE, scope);
+        heapSegment = MemorySegment.ofArray(new float[ELEM_SIZE]);
```
Comment (PaulSandoz) [inline test/micro/org/openjdk/bench/jdk/incubator/foreign/LoopOverSlice.java]: The heap `MemorySegment` wraps `float[]` but is accessed as if `int[]`, same bit size, but I suspect it was not intentional?

**Review by PaulSandoz — APPROVED** [review]

**Review by PaulSandoz — APPROVED** [review]

**Pull request closed without merging**